Fail bootstrap if ip mapping is missing during double write stage

gleb-cloudius opened the pull request:
If a node restart just before it stores bootstrapping node's IP it will
not have ID to IP mapping for bootstrapping node which may cause failure
on a write path. Detect this and fail bootstrapping if it happens.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

#### Build Details:
- Duration: 1 hr 25 min
- Builder: spider4.cloudius-systems.com

kbr-scylla commented:
@yaronkaikov @benipeled what happened in this CI run? It looks like no tests were executed, and artifacts contain only some coredump file (from jenkins?)

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9271/consoleText

```
client_loop: send disconnect: Broken pipe
[2024-05-28T09:43:20.380Z] fetch-pack: unexpected disconnect while reading sideband packet
[2024-05-28T09:43:20.380Z] fatal: early EOF
[2024-05-28T09:43:20.380Z] fatal: fetch-pack: invalid index-pack output
[2024-05-28T09:43:20.381Z] 
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2846)
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2185)
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:635)
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:871)
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:170)
[2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
[2024-05-28T09:43:20.381Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
[2024-05-28T09:43:20.381Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2024-05-28T09:43:20.382Z] 	at hudson.remoting.Request$2.run(Request.java:377)
[2024-05-28T09:43:20.382Z] 	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
[2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[2024-05-28T09:43:20.382Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:137)
[2024-05-28T09:43:20.382Z] 	at java.base/java.lang.Thread.run(Thread.java:829)
[2024-05-28T09:43:20.383Z] 	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from 5.29.124.170/5.29.124.170:37868
[2024-05-28T09:43:20.383Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1787)
[2024-05-28T09:43:20.383Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
[2024-05-28T09:43:20.383Z] 		at hudson.remoting.Channel.call(Channel.java:1003)
[2024-05-28T09:43:20.383Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:153)
[2024-05-28T09:43:20.383Z] 		at jdk.internal.reflect.GeneratedMethodAccessor945.invoke(Unknown Source)
[2024-05-28T09:43:20.383Z] 		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2024-05-28T09:43:20.384Z] 		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:138)
[2024-05-28T09:43:20.384Z] 		at com.sun.proxy.$Proxy857.execute(Unknown Source)
[2024-05-28T09:43:20.384Z] 		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1222)
[2024-05-28T09:43:20.384Z] 		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1305)
[2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:136)
[2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:101)
[2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:88)
[2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[2024-05-28T09:43:20.385Z] 		at java.base/java.lang.Thread.run(Thread.java:829)
[2024-05-28T09:43:20.385Z] ERROR: Error cloning remote repo 'origin'
[2024-05-28T09:43:20.385Z] Retrying after 10 seconds
[2024-05-28T09:43:19.829Z] ERROR: Timeout after 60 minutes
```

Anyway restarting

yaronkaikov commented:
> @yaronkaikov @benipeled what happened in this CI run? It looks like no tests were executed, and artifacts contain only some coredump file (from jenkins?)
> 
> https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9271/consoleText
> 
> ```
> client_loop: send disconnect: Broken pipe
> [2024-05-28T09:43:20.380Z] fetch-pack: unexpected disconnect while reading sideband packet
> [2024-05-28T09:43:20.380Z] fatal: early EOF
> [2024-05-28T09:43:20.380Z] fatal: fetch-pack: invalid index-pack output
> [2024-05-28T09:43:20.381Z] 
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2846)
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2185)
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:635)
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:871)
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:170)
> [2024-05-28T09:43:20.381Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
> [2024-05-28T09:43:20.381Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
> [2024-05-28T09:43:20.381Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
> [2024-05-28T09:43:20.382Z] 	at hudson.remoting.Request$2.run(Request.java:377)
> [2024-05-28T09:43:20.382Z] 	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
> [2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> [2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> [2024-05-28T09:43:20.382Z] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> [2024-05-28T09:43:20.382Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:137)
> [2024-05-28T09:43:20.382Z] 	at java.base/java.lang.Thread.run(Thread.java:829)
> [2024-05-28T09:43:20.383Z] 	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from 5.29.124.170/5.29.124.170:37868
> [2024-05-28T09:43:20.383Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1787)
> [2024-05-28T09:43:20.383Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
> [2024-05-28T09:43:20.383Z] 		at hudson.remoting.Channel.call(Channel.java:1003)
> [2024-05-28T09:43:20.383Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:153)
> [2024-05-28T09:43:20.383Z] 		at jdk.internal.reflect.GeneratedMethodAccessor945.invoke(Unknown Source)
> [2024-05-28T09:43:20.383Z] 		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> [2024-05-28T09:43:20.384Z] 		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> [2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:138)
> [2024-05-28T09:43:20.384Z] 		at com.sun.proxy.$Proxy857.execute(Unknown Source)
> [2024-05-28T09:43:20.384Z] 		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1222)
> [2024-05-28T09:43:20.384Z] 		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1305)
> [2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:136)
> [2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:101)
> [2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:88)
> [2024-05-28T09:43:20.384Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
> [2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
> [2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> [2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> [2024-05-28T09:43:20.385Z] 		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> [2024-05-28T09:43:20.385Z] 		at java.base/java.lang.Thread.run(Thread.java:829)
> [2024-05-28T09:43:20.385Z] ERROR: Error cloning remote repo 'origin'
> [2024-05-28T09:43:20.385Z] Retrying after 10 seconds
> [2024-05-28T09:43:19.829Z] ERROR: Timeout after 60 minutes
> ```
> 
> Anyway restarting

Looks like we failed to checkout

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9273/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9273/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_ip_mappings
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9273/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9804/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1780/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9273/)
#### Failed Tests (1/30804):
- [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9273/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
#### Build Details:
- Duration: 5 hr 3 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9329/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9329/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_ip_mappings
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9329/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1805/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9829/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9329/)
#### Build Details:
- Duration: 5 hr 54 min
- Builder: i-02390428984fe5c1f (m5ad.12xlarge)

kbr-scylla commented:
> 🔴 CI State: FAILURE
> ...
> [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9273/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)

@gleb-cloudius what was this failure? Is this an unrelated issue (in which case a reference to GH issue is needed) or something that was fixed in later version of PR (I see you force-pushed but did not comment)?

gleb-cloudius commented:
> > 🔴 CI State: FAILURE
> > ...
> > [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9273/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
> 
> @gleb-cloudius what was this failure? Is this an unrelated issue (in which case a reference to GH issue is needed) or something that was fixed in later version of PR (I see you force-pushed but did not comment)?

I forced pushed because I addressed comments. The failure is unrelated.

kbr-scylla commented:
Then please post a reference to the relevant GH issue. If it doesn't exist yet, create it.
(We have this policy to prevent introduction of flakiness -- reference to existing issue serves as evidence that it's unrelated)

gleb-cloudius commented:
> > 🔴 CI State: FAILURE
> > ...
> > [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9273/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
> 
> @gleb-cloudius what was this failure? 

https://github.com/scylladb/scylladb/issues/18904

patjed41 commented on test/topology_custom/test_ip_mappings.py:
```diff
@@ -0,0 +1,52 @@
+    async def worker():
+        logger.info("Worker started")
+        while not stop_event.is_set():
+            for i in range(100):
```
> It just provides us with a number to use as a key/value.

Right, the loop is fine.

patjed41 commented on test/topology_custom/test_ip_mappings.py:
```diff
@@ -0,0 +1,52 @@
+            await asyncio.sleep(0.1)
+        logger.info("Worker stopped")
+
+    await manager.server_start(server_a.server_id)
+    await manager.driver_connect()
+
+    worker_task = asyncio.create_task(worker())
```
If I'm not wrong, we should always wait for CQL before querying the restarted servers. Tests that don't it can be flaky and fail, for example, 1/1000 times. I wouldn't risk it here.

Maybe I'm wrong, and we don't need it here. @kbr-scylla knows this stuff well.

kbr-scylla replied:
Agreed, better to be safe now than sorry later. We know from experience that `wait_for_cql_and_get_hosts` prevents certain type of flakiness

gleb-cloudius replied:
Guys do we really have nothing else to do instead of nit picking on a simple test? There are no comments here that warrant me wasting half a day on a context switch to this old forgotten branch. The test run at least 500 times without `wait_for_cql_and_get_hosts` so without a definite evidence that it is needed here why would I waste time adding it? Rest of the comment are of the same importance.  If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway. Hanging for an hour in case of a failure is a problem worth fixing, not this.

kbr-scylla replied:
> If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway.

It will also sleep for 20 seconds unnecessarily in successful runs. Slows down CI runs for everyone for no reason. Multiply by all CI runs we're doing and that's a lot of time wasted for everyone.

kbr-scylla replied:
> There are no comments here that warrant me wasting half a day on a context switch to this old forgotten branch. 

It shouldn't take more than 20 minutes to address these comments. And you can do it in the morning the next day before daily to avoid disruptive context switching.

> The test run at least 500 times without wait_for_cql_and_get_hosts so without a definite evidence that it is needed here why would I waste time adding it?

Fine. Maybe it's not needed anymore with never python-driver versions. We'll find out.

gleb-cloudius replied:
> > If the test fails in 20 seconds it is not the end for the world since it should 100% pass anyway.
> 
> It will also sleep for 20 seconds unnecessarily in successful runs. Slows down CI runs for everyone for no reason. Multiply by all CI runs we're doing and that's a lot of time wasted for everyone.

The test runs traffic for 20 seconds. What else do you propose to do? If you have something else in mind send the patch.

kbr-scylla replied:
Run traffic for 5 seconds? Or just do a pre-determined number of writes? I would expect that the first couple of writes would already reproduce the issue since every write needs the mapping.  Is running 20 seconds of workload really needed to reproduce the issue reliably?

gleb-cloudius replied:
Any of that is not less arbitrary but now the review comes down to bike shedding about 20 vs 5, while the patch could have been in the next for a couple hours already.

kbr-scylla replied:
Our goal is not to push as many patches as we can in shortest amount of time.

There is clear value in keeping tests short as possible. Best would be if we could make this test not time dependent (sleep for X) but just do e.g. 100 writes and if there's a 99% chance that one of these writes will hit the no mapping issue then it's sufficient, and the writing part will take few milliseconds in most runs.

gleb-cloudius replied:
> Our goal is not to push as many patches as we can in shortest amount of time.

There should be a balance. Bike shedding 5s vs 20s is just not it given that we do want some timeout anyway. It saturates developers time with unimportant stuff. I can nitpick any non trivial patch for eternity. There is no pride in it.

kbr-scylla replied:
Enough with the 5s vs 20s.

About this:
> Best would be if we could make this test not time dependent (sleep for X) but just do e.g. 100 writes and if there's a 99% chance that one of these writes will hit the no mapping issue then it's sufficient, and the writing part will take few milliseconds in most runs.

You still did not explain why you decided to go with the solution of "run workload for X seconds in background and then check" instead of "do N inserts in foreground" -- which is both faster and results in simpler code.

I checked the second version, works reliably (100/100 runs) with N = 10000. Compare:
```python
    stop_event = asyncio.Event()
    async def worker():
        logger.info("Worker started")
        while not stop_event.is_set():
            for i in range(100):
                await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
                response = await manager.cql.run_async(f"SELECT * FROM test.test WHERE a = {i}")
                assert response[0].b == i
            await asyncio.sleep(0.1)
        logger.info("Worker stopped")

    await manager.server_start(server_a.server_id)
    await manager.driver_connect()

    worker_task = asyncio.create_task(worker())

    await asyncio.sleep(20)
    stop_event.set()
    await worker_task
```

with:
```python
    await manager.server_start(server_a.server_id)
    await manager.driver_connect()

    for i in range(10000):
        await manager.cql.run_async(f"INSERT INTO test.test (a, b) VALUES ({i}, {i})")
        response = await manager.cql.run_async(f"SELECT * FROM test.test WHERE a = {i}")
        assert response[0].b == i
```

The first one is unnecessarily complex and makes the test run 20 seconds longer.

It's not a nitpick, it would be a significant simplification to the code and improvement for our test suite performance.

(And do you want me to point you to some of reviews you did on my code because you had a simpler design in mind? And I didn't complain and call your suggestions "nitpicks")

gleb-cloudius replied:
> You still did not explain why you decided to go with the solution of "run workload for X seconds in background and then check" instead of "do N inserts in foreground" -- which is both faster and results in simpler code.

Because I did not wrote the test and I do not strive for perfection in a test for a very specific problem. Would I wrote it differently? Probably. But why would I write a new test if one is a;ready exist.

and for me if the existing test do the job is good enough. I do not strive for code perfection in a test for a very specific problem.

> 
> (And do you want me to point you to some of reviews you did on my code because you had a simpler design in mind? And I didn't complain and call your suggestions "nitpicks")

About a test? Please show me.

kbr-scylla replied:
We agreed on daily to address this in a follow-up.

kbr-scylla commented on test/topology_custom/test_ip_mappings.py:
```diff
@@ -0,0 +1,55 @@
+logger = logging.getLogger(__name__)
+
+@pytest.mark.asyncio
+async def test_broken_bootstrap(manager: ManagerClient):
```
Please add a comment that the test is for reproducing scylladb/scylladb#18676

kbr-scylla commented on test/topology_custom/test_ip_mappings.py:
```diff
@@ -0,0 +1,55 @@
+
+    worker_task = asyncio.create_task(worker())
+
+    await asyncio.sleep(20)
```
And because this is a background task, if the test fails after 1 second of writing, we'll only learn about the failure after 20 seconds, the test will unnecessarily sleep for 19 seconds.

kbr-scylla commented:
Congratulations on your first closed pull request @gleb-cloudius!
